### DISCUSSION TITLE
Discussion around tests

### DISCUSSION THREAD

**Discussion by MerlinEgalite** [morpho-labs/data-structures#122] [Announcements]
## Context

Below is a part of the comment from @makcandrov about tests on this PR: https://github.com/morpho-dao/morpho-data-structures/pull/117. I'm creating this discussion to avoid polluting the PR.

## Message 

While refreshing this library, I came across many incoherent things and remnants of old designs and testing tricks. In order to refresh and harmonize all the files from this repository, I think that some other changes are needed, but I believe that they deserve separate PRs. They are:

- [ ] All the testing files use the extension `.t.sol`, which we don't use anymore in recent repositories.  
- [ ] All the files use the old syntax of imports (`import "./TestCommonHeapOrdering.t.sol";` instead of `import {TestCommonHeapOrdering} from "./TestCommonHeapOrdering.t.sol";`).
**EDIT:** issue: #125 - pending fix: #126
- [ ] Maybe it would be better to gather all the invariant tests files in a separate folder under `./test/`.
**EDIT:** pending fix: #130 
- [ ] Because we are testing libraries, we cannot make a relevant gas report if we directly call the internal functions on the heap. That's why some tests use a "concrete heap" contract (see [`ConcreteHeapOrdering.sol`](https://github.com/morpho-dao/morpho-data-structures/blob/main/test/helpers/ConcreteHeapOrdering.sol)) that is essentially the "contract version" of the library. However, not all tests use this (see [`TestHeap.t.sol`](https://github.com/morpho-dao/morpho-data-structures/blob/main/test/TestHeap.t.sol)), and some of them add functions that are not in the library in the first place (see `accountsValue`, `accountsId` and `indexOf` from [`ConcreteHeapOrdering.sol`](https://github.com/morpho-dao/morpho-data-structures/blob/main/test/helpers/ConcreteHeapOrdering.sol) and [`ConcreteThreeHeapOrdering.sol`](https://github.com/morpho-dao/morpho-data-structures/blob/main/test/helpers/ConcreteThreeHeapOrdering.sol)). In my opinion, all tests should use a "contract version" of the library but with strictly the same functions.
**EDIT:** issue: #127 - pending partial fix: #131
- [ ] The testing suite generally lacks fuzzing.
**EDIT:** issue: #127 - pending partial fix: #131
- [ ] Many tests check that the storage is exactly what we expect after a series of operations (see [`testRemoveShiftUp`](https://github.com/morpho-dao/morpho-data-structures/blob/ba30cb23052c14dfb399ce87109d06fd868022e1/test/TestHeap.t.sol#L294-L319)). I don't think these are good tests because they are highly dependent on the implementation of the heap. For the same series of operations, many different orders in the list could be correct. For example if we insert the values `0`, `1`, and `2` in the heap, the lists `[2, 0, 1]` and `[2, 1, 0]` would be correct. However, the testing suite would fail in one of these situations.
**EDIT:** issue: #127 - pending partial fix: #131

**Comment by MerlinEgalite** [comment]
> All the testing files use the extension .t.sol, which we don't use anymore in recent repositories.

Maybe we should have kept the `.t.sol` in other repositories to follow [foundry best practices](https://book.getfoundry.sh/tutorials/best-practices?highlight=best#general-test-guidance).

> All th files use the old syntax of imports (import "./TestCommonHeapOrdering.t.sol"; instead of import {TestCommonHeapOrdering} from "./TestCommonHeapOrdering.t.sol";)

I agree we should change it.

> Maybe it would be better to gather all the invariant tests files in a separate folder under ./test/.

I don't have a strong opinion on this so I'm ok.

>  Because we are testing libraries, ...

I agree with this comment

> The testing suite generally lacks fuzzing.

Agree

> Many tests check that the storage is exactly what we expect ...

I'm not sure about what you're suggesting here? Removing those tests?

**Reply by makcandrov** [reply]
> Maybe we should have kept the .t.sol in other repositories to follow [foundry best practices](https://book.getfoundry.sh/tutorials/best-practices?highlight=best#general-test-guidance).

I'll let you decide then

> I don't have a strong opinion on this so I'm ok.

I think that the more data structures we have in this repository, the messier the test folder will become. Separating them by test type or by data structure could be a good idea.

> I'm not sure about what you're suggesting here? Removing those tests?

Not necessarily remove them, but rather make them less dependent on the specific implementation details. For example, instead of checking that the storage is exactly `[2, 0, 1]` when inserting the numbers `0`, `1`, and `2`, we could check that the root is `2`, one of its children is `0`, and the other is `1`.
</br>
About the other points, I might wait for #119 to be merged to create the issues. Since it changes a lot of things, we might have a clearer picture afterward.

**Reply by MathisGD** [reply]
About the `.t.sol` extension: i think that it's useless to have the "Test" prefix in the files name if we have this (which I'm in favor)

**Reply by pakim249CAL** [reply]
>  Maybe it would be better to gather all the invariant tests files in a separate folder under ./test/.

I think this is a good practice to adopt especially for public users that are just viewing the repo to be able to understand more quickly how tests are structured.

>> Many tests check that the storage is exactly what we expect ...

> I'm not sure about what you're suggesting here? Removing those tests?

I believe what is being suggested is to define tests based on how the data structure should behave rather than how it is implemented. There can be multiple orders of variables that would make a "correct" heap implementation, so tests should be structured to accept all good solutions.

I agree with all of Merlin's other points.

**Reply by QGarchery** [reply]
I agree with @pakim249CAL but I also don't like the `.t.sol` extension, because this convention does not work well with contract names. For example a file `Heap.t.sol` might have `Heap` as contract name and then it conflicts with the file `Heap.sol` which rightly contains the contract with the same name

**Reply by MerlinEgalite** [reply]
You mean it's an issue when you're doing a file or variable research @QGarchery right?

**Reply by QGarchery** [reply]
I was thinking about having a solidity file name that is not the same as the contract declaration it contains. So I like that a file is named `TestHeap.sol` and the contract declaration is `contract TestHeap {...`. With the `.t.sol` convention, the file name would be `Heap.t.sol` and the contract declaration would be `contract Heap {...` which is awkward because there might be another contract with that name

**Reply by MathisGD** [reply]
>I was thinking about having a solidity file name that is not the same as the contract declaration it contains.

This is the "standard" in forge based repos (Heap.t.sol and contract TestHeap)

**Comment by makcandrov** [comment]
@QGarchery you may have an interesting opinion on the last point

**Reply by makcandrov** [reply]
btw I created the issue #127 to address the points 4 & 5. I am waiting for other opinions before addressing the last point

**Reply by QGarchery** [reply]
Not really, I completely agree that test should not be dependent on the implementation. The fact that they are is more a product of lack of time than anything else
